lib/Sema/TypeCheckType.cpp

```diff
@@ -1061,6 +1061,14 @@ static Type diagnoseUnknownType(TypeResolution resolution,
         // type. Fix this by replacing 'Self' with the nominal type name.
         assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
 
+        bool insideClass = nominalDC->getSelfClassDecl() != nullptr;
+        if (!insideClass || options.isAnyExpr()) {
```

Comment:
I wonder if there's a way to combine this with the existing code elsewhere in the file -- search for this:

```
  // Dynamic 'Self' in the result type of a function body.
  if (options.getBaseContext() == TypeResolverContext::DynamicSelfResult &&
      comp->getIdentifier() == ctx.Id_Self) {
```

Reply:
I see what you mean. Is this the rewrite you were talking about?